### PR TITLE
OSSL_FN: Add the X9.31 prime functions [pending #32485]

### PR DESCRIPTION
**Pull request by levitte**
Port the BN X9.31 prime machinery (bn_x931p.c) to OSSL_FN.  This adds the OSSL_FN counterparts of the X9.31 parameter-pair and prime derivation functions, all with sizing / testing per the OSSL_FN porting rules.  The BN_X931_* functions are untouched; RSA X9.31 keygen switches to these later.

Functions added:

- OSSL_FN_X931_generate_Xpq() (+ _ctx_size) — X9.31 parameter pair generation (peer of BN_X931_generate_Xpq()).
- OSSL_FN_X931_derive_prime() (+ _ctx_size) — prime derivation from Xp1 / Xp2 / Xp (peer of BN_X931_derive_prime_ex()), with the static bn_x931_derive_pi() helper carried along as file-static ossl_fn_x931_derive_pi().
- OSSL_FN_X931_generate_prime() (+ _ctx_size) — full X9.31 prime generation (peer of BN_X931_generate_prime_ex()).

The derivation is cross-checked against BN_X931_derive_prime_ex() in fn_cross_bn_test.

Please see the individual commit messages for details.

Related-to: doc/designs/fixed-size-large-numbers.md
Issue: https://github.com/openssl/project/issues/1654
Link: https://github.com/openssl/openssl/pull/32485

Assisted-by: Pi:moonshotai/kimi-k3

### CODE REVIEW

**Comment by levitte** [comment]
This PR depends on other PRs (see PR title); only the last 4 commits belong with this PR.  I recommend reviewing the other PRs first.  When reviewing this PR, it may be wise to review each commit separately, to avoid confusion.

**Comment by levitte** [comment]
The fixup that was pushed can be made simpler if https://github.com/openssl/openssl/pull/32725 is merged.

File: crypto/fn/fn_x931p.c [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        goto err;
+    if (!OSSL_FN_mod_mul(p, p, p2, p1p2, ctx))
+        goto err;
+    if (!OSSL_FN_mod_inverse(t, p1, p2, ctx))
+        goto err;
+    if (!OSSL_FN_mod_mul(t, t, p1, p1p2, ctx))
+        goto err;
+    if (!OSSL_FN_mod_sub(p, p, t, p1p2, ctx))
+        goto err;
+
```
Comment (mattcaswell) [inline crypto/fn/fn_x931p.c:238]: At this point in `BN_X931_derive_prime_ex` we see this code:

``` C
    /* p now equals Rp */

    if (!BN_mod_sub(p, p, Xp, p1p2, ctx))
        goto err;

    if (!BN_add(p, p, Xp))
        goto err;
```

This seems to be entirely missing here!

The fact that the tests pass with this problem suggests a coverage problem in the tests. Claude tells me that the tests pass with or without the fix for this.

File: crypto/fn/fn_x931p.c [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    if (Xp == NULL || Xp->dsize == 0)
+        return 0;
+
+    return OSSL_FN_CTX_size(1, 1, (size_t)Xp->dsize);
+}
+
+/*
+ * Estimate the arena payload for OSSL_FN_X931_derive_prime().  The outer
+ * frame holds five numbers (p1, p2 of n limbs; p1p2 of 2n limbs; t and pm1
+ * of n limbs), so the scratch budget is 5n limbs over 5 numbers.  The
```
Comment (mattcaswell) [inline crypto/fn/fn_x931p.c]: Surely that adds up to 6n limbs?

File: crypto/fn/fn_x931p.c [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    if ((token = OSSL_FN_CTX_start(ctx)) == NULL)
+        return 0;
+
+    if (p1 == NULL && (p1 = OSSL_FN_CTX_get_limbs(ctx, (size_t)Xp->dsize)) == NULL)
+        goto err;
+    if (p2 == NULL && (p2 = OSSL_FN_CTX_get_limbs(ctx, (size_t)Xp->dsize)) == NULL)
+        goto err;
+
+    p1p2 = OSSL_FN_CTX_get_limbs(ctx, 2 * (size_t)Xp->dsize);
+    t = OSSL_FN_CTX_get_limbs(ctx, (size_t)Xp->dsize);
```
Comment (mattcaswell) [inline crypto/fn/fn_x931p.c]: `t` is reduced modulo `p1p2` below which implies it needs the same double width as `p1p2`

File: crypto/fn/fn_x931p.c [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+    /* Compute Rp = (p2^-1 mod p1) * p2 - (p1^-1 mod p2) * p1 (mod p1p2),
+     * then Yp0 = Xp + Rp - Xp modulo p1p2.  The products are reduced
+     * against p1p2 rather than p1 or p2, so the CRT difference survives.
+     * The sign fix-up of the BN counterpart collapses into the modular
+     * reduction.
+     */
+    if (!OSSL_FN_mod_inverse(p, p2, p1, ctx))
+        goto err;
+    if (!OSSL_FN_mod_mul(p, p, p2, p1p2, ctx))
```
Comment (mattcaswell) [inline crypto/fn/fn_x931p.c:230]: So `p` needs to be double width right? But the comment in the header for this function says:

> Provided operands must carry the main width (@p Xp's width): @p Xp1, @p Xp2, and any non-NULL @p p1 / @p p2 all do.

This doesn't actually mention `p` at all explicitly but the implication is that all operands (which I assume includes `p`) must be the same width as `Xp`.

File: include/crypto/fn.h [hunk 5]
```diff
@@ -1454,6 +1521,170 @@ int OSSL_FN_from_mont(OSSL_FN *r, const OSSL_FN *a,
...
+int OSSL_FN_X931_derive_prime(OSSL_FN *p, OSSL_FN *p1, OSSL_FN *p2,
+    const OSSL_FN *Xp, const OSSL_FN *Xp1, const OSSL_FN *Xp2,
+    const OSSL_FN *e, OSSL_FN_CTX *ctx, BN_GENCB *cb,
+    OSSL_LIB_CTX *libctx);
+
+/**
+ * Generate an X9.31 prime @p p using random Xp1 and Xp2 parameters, then
+ * defer to OSSL_FN_X931_derive_prime().  Any non-NULL @p p1, @p p2, @p Xp1
+ * and @p Xp2 receive the intermediate values.  All operands carry @p Xp's
+ * width, which must be at least 101 bits wide for the parameter draws.
```
Comment (mattcaswell) [inline include/crypto/fn.h]: This is not true for `p`. See my comments in `OSSL_FN_X931_derive_prime`

File: test/fn_api_test.c [hunk 6]
```diff
@@ -891,6 +892,371 @@ static int test_introspection(void)
...
+err:
+    OSSL_FN_free(a);
+    OSSL_FN_free(z_empty);
+    return ret;
+}
+
+/*-
+ * Focused tests for ossl_fn_check_prime() / ossl_fn_check_generated_prime()
+ * and the underlying ossl_fn_miller_rabin_is_prime().  Small primes and
+ * composites are checked by verdict; the arena is sized generously and the
```
Comment (mattcaswell) [inline test/fn_api_test.c]: Why are we doing that? Surely we want to be testing the ctx_size functions?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/fn_api_test.c (hunk 6 above)
Comment (levitte) [reply test/fn_api_test.c]: These and other changes that don't belong in this PR were experimentation to figure out some edge cases...  should never have slipped into this PR.  Undoing them now.

File: crypto/fn/fn_x931p.c (hunk 1 above)
Comment (levitte) [reply crypto/fn/fn_x931p.c:238]: Good catch.  mod_sub + add restored.

File: crypto/fn/fn_x931p.c (hunk 2 above)
Comment (levitte) [reply crypto/fn/fn_x931p.c]: Yeah, and with t and pm1 going double width per your other comments, it's actually 8n.

I also found that the scratch budget allocated hard-coded numbers (incidently, *5 * (size_t)Xp->dsize*), which turns out not to be a good choice.  It's now replaced with a spelling out of the operand widths (p1 + p2 + p1p2 + t + pm1).

File: crypto/fn/fn_x931p.c (hunk 3 above)
Comment (levitte) [reply crypto/fn/fn_x931p.c]: Fixed

File: crypto/fn/fn_x931p.c (hunk 4 above)
Comment (levitte) [reply crypto/fn/fn_x931p.c:230]: True.  Fixed.

File: include/crypto/fn.h (hunk 5 above)
Comment (levitte) [reply include/crypto/fn.h]: Fixed

File: crypto/fn/fn_x931p.c [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        ossl_fn_ctx_add_size(ossl_fn_ctx_add_size(mul, inv),
+            ossl_fn_ctx_add_size(mm, ms)));
+    return ossl_fn_ctx_add_size(own, loop);
+}
+
+size_t OSSL_FN_X931_generate_prime_ctx_size(const OSSL_FN *p,
+    const OSSL_FN *p1, const OSSL_FN *p2, const OSSL_FN *Xp,
+    const OSSL_FN *e)
+{
+    return OSSL_FN_X931_derive_prime_ctx_size(p, p1, p2, Xp, NULL, NULL,
```
Comment (mattcaswell) [inline crypto/fn/fn_x931p.c]: `OSSL_FN_X931_generate_prime` creates a frame and four numbers. Shouldn't that be reflected here?

File: crypto/fn/fn_x931p.c [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    if (!OSSL_FN_mod_mul(p, p, p2, p1p2, ctx))
+        goto err;
+    if (!OSSL_FN_mod_inverse(t, p1, p2, ctx))
+        goto err;
+    if (!OSSL_FN_mod_mul(t, t, p1, p1p2, ctx))
+        goto err;
+    if (!OSSL_FN_mod_sub(p, p, t, p1p2, ctx))
+        goto err;
+
+    /* p now equals Rp; normalize to Yp0 = Xp + Rp - Xp (mod p1p2). */
```
Comment (mattcaswell) [inline crypto/fn/fn_x931p.c]: This is really:


```suggestion
    /* p now equals Rp; normalize to Yp0 = Xp + ((Rp - Xp) mod p1p2). */
```

File: test/fn_api_test.c
Comment (mattcaswell) [inline test/fn_api_test.c:1]: Claude suggests this diff to act as a regression test for the bug you previously fixed:

``` C
diff --git a/test/fn_api_test.c b/test/fn_api_test.c
index 8f296954a0..2cdb446ee2 100644
--- a/test/fn_api_test.c
+++ b/test/fn_api_test.c
@@ -6094,9 +6094,11 @@ static int test_x931_derive_prime(void)
 
     /*
      * Small auxiliary primes keep Yp0 + k*p1*p2 comfortably inside one
-     * limb even on 32-bit builds.
+     * limb even on 32-bit builds.  Xp sits well above p1 * p2 (~2^21
+     * here), matching how X9.31 is used: Xp then determines the derived
+     * prime's magnitude via the Yp0 = Xp + ((Rp - Xp) mod p1p2) fold.
      */
-    if (!TEST_true(OSSL_FN_set_word(Xp, 0xBEEFUL))
+    if (!TEST_true(OSSL_FN_set_word(Xp, 0x0BEEFBEFUL))
         || !TEST_true(OSSL_FN_set_word(Xp1, 1021))
         || !TEST_true(OSSL_FN_set_word(Xp2, 1031))
         || !TEST_true(OSSL_FN_set_word(e, 65537))
@@ -6122,6 +6124,16 @@ static int test_x931_derive_prime(void)
     if (!TEST_int_eq(ossl_fn_check_prime(p, 0, ctx, 1, NULL, NULL), 1))
         goto err;
 
+    /*
+     * Yp0 is the smallest integer >= Xp congruent to Rp modulo p1 * p2,
+     * and the search only ever adds p1 * p2, so the result is >= Xp.
+     * The CRT congruences below hold for Rp as well as for Yp0, so this
+     * is what separates the two: without the Xp fold the result stays
+     * below p1 * p2 and so below Xp.
+     */
+    if (!TEST_int_ge(OSSL_FN_cmp(p, Xp), 0))
+        goto err;
+
     /* CRT congruences: p == 1 (mod p1), p == -1 (mod p2). */
     if (!TEST_true(OSSL_FN_div(NULL, r, p, p1, ctx))
         || !TEST_true(OSSL_FN_is_word(r, 1)))
@@ -6177,9 +6189,14 @@ static int test_x931_generate_prime(void)
     OSSL_FN *Xp = NULL, *e = NULL;
     OSSL_FN_CTX *ctx = NULL;
     size_t size = 0;
-    /* Xp1 / Xp2 are drawn as 101-bit numbers; the buffers must fit them,
-     * while p must be double-wide, as derive_prime requires. */
-    size_t limbs = (101 + OSSL_FN_BITS - 1) / OSSL_FN_BITS;
+    /*
+     * Xp1 / Xp2 are drawn as 101-bit numbers, so p1 * p2 is a little
+     * over 202 bits.  Take the main width wider than that, so that Xp
+     * exceeds p1 * p2 as it does in X9.31's actual use (Xp is half the
+     * modulus width there); p carries double the main width, as
+     * derive_prime requires.
+     */
+    size_t limbs = (256 + OSSL_FN_BITS - 1) / OSSL_FN_BITS;
 
     if (!TEST_ptr(p = OSSL_FN_new_limbs(2 * limbs))
         || !TEST_ptr(p1 = OSSL_FN_new_limbs(limbs))
@@ -6190,7 +6207,9 @@ static int test_x931_generate_prime(void)
         || !TEST_ptr(e = OSSL_FN_new_limbs(1)))
         goto err;
 
+    /* A 255-bit Xp, comfortably above p1 * p2. */
     if (!TEST_true(OSSL_FN_set_word(Xp, 0xDEADBEEFUL))
+        || !TEST_true(OSSL_FN_set_bit(Xp, 254))
         || !TEST_true(OSSL_FN_set_word(e, 65537)))
         goto err;
 
@@ -6207,6 +6226,14 @@ static int test_x931_generate_prime(void)
     if (!TEST_int_eq(ossl_fn_check_prime(p, 0, ctx, 1, NULL, NULL), 1))
         goto err;
 
+    /*
+     * The generated prime takes its magnitude from Xp, not from the
+     * auxiliary primes: without the Yp0 = Xp + ((Rp - Xp) mod p1p2)
+     * fold the result would be a ~202-bit residue instead.
+     */
+    if (!TEST_int_ge(OSSL_FN_cmp(p, Xp), 0))
+        goto err;
+
     /* Xp1 / Xp2 were drawn as 101-bit numbers with the top bit set. */
     if (!TEST_size_t_eq(OSSL_FN_num_bits(Xp1), 101)
         || !TEST_size_t_eq(OSSL_FN_num_bits(Xp2), 101))
diff --git a/test/fn_cross_bn_test.c b/test/fn_cross_bn_test.c
index 277ae9724a..a6d56ce8e3 100644
--- a/test/fn_cross_bn_test.c
+++ b/test/fn_cross_bn_test.c
@@ -125,9 +125,12 @@ static int test_x931_derive_prime_cross_bn(void)
     /*
      * Small auxiliary parameters keep the derived prime comfortably
      * inside one limb even on 32-bit builds, same as the corresponding
-     * fn_api_test coverage.
+     * fn_api_test coverage.  Xp is chosen well above p1 * p2 (~2^21 for
+     * these auxiliaries), the regime X9.31 is actually used in: the
+     * Xp fold then sets the derived prime's magnitude, so dropping it
+     * changes the result rather than being absorbed by the search loop.
      */
-    if (!TEST_true(BN_set_word(bXp, 0xBEEF))
+    if (!TEST_true(BN_set_word(bXp, 0x0BEEFBEF))
         || !TEST_true(BN_set_word(bXp1, 1021))
         || !TEST_true(BN_set_word(bXp2, 1031))
         || !TEST_true(BN_set_word(be, 65537)))
```

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/fn/fn_x931p.c (hunk 7 above)
Comment (levitte) [reply crypto/fn/fn_x931p.c]: Yes it should.  Fixed.

File: crypto/fn/fn_x931p.c (hunk 8 above)
Comment (levitte) [reply crypto/fn/fn_x931p.c]: Taken, thank you

File: test/fn_api_test.c
Comment (levitte) [reply test/fn_api_test.c:1]: Okie, included.  Thank you

**Review by mattcaswell — APPROVED** [review]

**Review by n13l — APPROVED** [review]
